app/utils/deep_link/index.ts

```diff
@@ -114,80 +123,95 @@ export async function handleDeepLink(deepLinkUrl: string, intlShape?: IntlShape,
+    serverUrl: string;
+    teamName: string;
+    postId: string;
+};
+const PERMALINK_PATH = `:serverUrl(.*)/:teamName(${TEAM_NAME_PATH_PATTERN})/pl/:postId(${ID_PATH_PATTERN})`;
+export const matchPermalinkDeeplink = match<PermalinkPathParams>(PERMALINK_PATH);
+
 export function parseDeepLink(deepLinkUrl: string): DeepLinkWithData {
     try {
-        const url = removeProtocol(decodeURIComponent(deepLinkUrl));
+        const url = removeProtocol(deepLinkUrl);
 
-        if (url.includes('../') || url.includes('/..')) {
-            return {type: DeepLink.Invalid, url: deepLinkUrl};
-        }
+        const channelMatch = matchChannelDeeplink(url);
+        if (channelMatch) {
+            const {params: {serverUrl, teamName, path, identifier}} = channelMatch;
 
-        let match = new RegExp('(.*)\\/([^\\/]+)\\/channels\\/(\\S+)').exec(url);
-        if (match) {
-            return {type: DeepLink.Channel, url: deepLinkUrl, data: {serverUrl: match[1], teamName: match[2], channelName: match[3]}};
-        }
+            if (path === 'channels') {
+                return {type: DeepLink.Channel, url: deepLinkUrl, data: {serverUrl, teamName, channelName: identifier}};
+            }
 
-        match = new RegExp('(.*)\\/([^\\/]+)\\/pl\\/(\\w+)').exec(url);
-        if (match) {
-            return {type: DeepLink.Permalink, url: deepLinkUrl, data: {serverUrl: match[1], teamName: match[2], postId: match[3]}};
-        }
+            if (path === 'messages') {
+                if (identifier.startsWith('@')) {
+                    return {type: DeepLink.DirectMessage, url: deepLinkUrl, data: {serverUrl, teamName, userName: identifier.substring(1)}};
+                }
 
-        match = new RegExp('(.*)\\/([^\\/]+)\\/messages\\/@(\\S+)').exec(url);
-        if (match) {
-            return {type: DeepLink.DirectMessage, url: deepLinkUrl, data: {serverUrl: match[1], teamName: match[2], userName: match[3]}};
+                return {type: DeepLink.GroupMessage, url: deepLinkUrl, data: {serverUrl, teamName, channelName: identifier}};
+            }
         }
 
-        match = new RegExp('(.*)\\/([^\\/]+)\\/messages\\/(\\S+)').exec(url);
-        if (match) {
-            return {type: DeepLink.GroupMessage, url: deepLinkUrl, data: {serverUrl: match[1], teamName: match[2], channelId: match[3]}};
+        const permalinkMatch = matchPermalinkDeeplink(url);
+        if (permalinkMatch) {
+            const {params: {serverUrl, teamName, postId}} = permalinkMatch;
+            return {type: DeepLink.Permalink, url: deepLinkUrl, data: {serverUrl, teamName, postId}};
         }
 
-        match = new RegExp('(.*)\\/plugins\\/([^\\/]+)\\/(\\S+)').exec(url);
-        if (match) {
-            return {type: DeepLink.Plugin, url: deepLinkUrl, data: {serverUrl: match[1], id: match[2], teamName: ''}};
+        const pluginMatch = match<{serverUrl: string; id: string; route?: string[]}>(`:serverUrl(.*)/plugins/:id(${PLUGIN_ID_PATH_PATTERN})/:route+`)(url);
```

Comment:
Ok this is still surprisingly tricky to get right. `match` internally compiles the whole path to a regexp and the addition of the `/:path+` component here suddenly makes that regexp polynomial in complexity and vulnerable to ReDoS. And not just that, but `:route+` also matches `../../../..` etc., which is the original issue we're attempting to fix here (yet in this case would only manifest once we implement MM-49846).

To address the ReDoS we could just limit the input length to something sane. The second part is tough because we don't know what plugin URLs are supposed to look like.

Reply:
Yes, the `/:route` param opens up a potential hole down the road if future codepaths do not further positively match/validate known plugin routes.

To address the ReDoS part, I've switched it to `/:route(.*)?` to just take it as one group as `:serverUrl` does.

As to the second part, IMO it is not feasible to deterministically match arbitrary plugin routes in `matchDeepLink`/`parseDeepLink` at this point in time without some sort of deeplink route registry that plugins could define—I only added `/:route` for future use that would require downstream matching, logic, and so on.

Because `case DeepLink.Plugin:` in handleDeepLink is currently a no-op, I don't see `/:route` as an actual hole, but rather like rough-in plumbing—e.g. a pipe with a removable cap on the end of it for expansion later on.

All of the other matched parameters that are presently acted upon in these flows are validated before any requests are made, and unless there are any other outstanding issues, I propose that we move this forward with the known requirements of handling `/:route` and `DeepLink.Plugin` properly in future development.

For tracking, we could split https://mattermost.atlassian.net/browse/MM-53902 into a `DeepLink.Plugin` version, or the validation requirements can be documented directly in https://mattermost.atlassian.net/browse/MM-49846.

Reply:
Are we actually using the plugins deeplinks? @larkox you would know more, but the only thing I can think of is the handling of a slash command response or interactive dialog.. am I correct?

Reply:
`handleDeepLink` is used both in commands "go to location" and `markdownLink` "onPress".

The main usage in plugins as of today is to handle OAuth. But the reality is that it never got implemented for V2 (https://mattermost.atlassian.net/browse/MM-49846).

So... should we actually use plugin deeplinks? Yes if we want to allow smooth oauth connection on mobile with the current plugin architecture. Are we actually using the plugins deeplinks? No.